### PR TITLE
Stream read bug fix

### PR DESCRIPTION
**Pull request by rccarper**
*Issue #, if available:*
Fixing the assumption that the specified content-length of the file being uploaded is the same size as the stream. The buffer we read into the from the specified input stream will now always be the exact required size for the last part, instead of always just the part size.

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: source/s3_auto_ranged_put.c [hunk 1]
```diff
@@ -345,9 +347,20 @@ static int s_s3_auto_ranged_put_next_request(
                     0,
                     AWS_S3_REQUEST_DESC_RECORD_RESPONSE_HEADERS);
 
-                aws_byte_buf_init(&request->request_body, meta_request->allocator, meta_request->part_size);
                 request->part_number = auto_ranged_put->threaded_next_request_data.next_part_number;
 
+                size_t request_body_size = meta_request->part_size;
+
+                if (request->part_number == auto_ranged_put->synced_data.total_num_parts) {
```
Comment (justinboswell) [inline source/s3_auto_ranged_put.c:354]: `/* last part, adjust size to match remaining content length */`

**Review by justinboswell — APPROVED** [review]
Fix-n-shipperooni

File: include/aws/s3/private/s3_auto_ranged_put.h [hunk 2]
```diff
@@ -33,6 +33,8 @@ struct aws_s3_auto_ranged_put {
     /* Useable after the Create Multipart Upload request succeeds. */
     struct aws_string *upload_id;
 
+    size_t content_length;
```
Comment (graebm) [inline include/aws/s3/private/s3_auto_ranged_put.h:36]: uint64_t?

File: source/s3_auto_ranged_put.c (hunk 1 above)
Comment (graebm) [inline source/s3_auto_ranged_put.c:354]: just double checking: part_number starts at 1 and not 0, right?

Comment (rccarper) [reply source/s3_auto_ranged_put.c:354]: Correct!  S3 part numbers start at 1.

**Review by graebm — APPROVED** [review]

**Pull request merged**